### PR TITLE
Fix various TODO items

### PR DESCRIPTION
**Pull request by gammazero**
- Republisher can get updates in retry loop
- Concurrent calls to Root.Close are safe
- Root.Close closess closes everything

### CODE REVIEW

File: dir.go [hunk 1]
```diff
@@ -122,8 +122,8 @@ func (d *Directory) localUpdate(c child) (*dag.ProtoNode, error) {
 		return nil, err
 	}
 
+	// A copy is needed here since nd is referenced by dagService.
```
Comment (schomatis) [inline dir.go:125]: Could you elaborate? I understand the `dagService` takes the node as (read-only) input and process it but as far as I know this happens sequentially: once we reach this line already `dagService` added the node and doesn't hold to the reference anymore.

File: dir.go [hunk 2]
```diff
@@ -405,7 +405,8 @@ func (d *Directory) sync() error {
 		}
 	}
 
-	// TODO: Should we clean the cache here?
+	// Clear entries cache
+	d.entriesCache = make(map[string]FSNode)
```
Comment (schomatis) [inline dir.go:409]: What happens if there is an error in the loop above and we already process half of this? (Maybe an error already signals corruption and the cache state is no longer important; I'm not sure what are the expectations here.)

File: file.go [hunk 3]
```diff
@@ -138,7 +139,6 @@ func (fi *File) Size() (int64, error) {
 }
 
 // GetNode returns the dag node associated with this file
-// TODO: Use this method and do not access the `nodeLock` directly anywhere else.
```
Comment (schomatis) [inline file.go:141]: :raised_hands:

File: file.go [hunk 4]
```diff
@@ -160,14 +160,15 @@ func (fi *File) Flush() error {
...
 
-	defer fd.Close()
-
-	return fd.Flush()
+	// Close calls fd.Flush() and will do a full sync since it was opened with
+	// Sync=true. So, there is not need to call fd.Flush()
+	return fd.Close()
 }
 
 func (fi *File) Sync() error {
```
Comment (schomatis) [inline file.go:168]: Who even calls this? Can't we just remove it?

File: file.go [hunk 5]
```diff
@@ -160,14 +160,15 @@ func (fi *File) Flush() error {
...
-	// just being able to take the writelock means the descriptor is synced
-	// TODO: Why?
+	// The descriptor lock is locked in File.Open and unlocked in
+	// fileDescriptor.Close. Taking the writelock ensures that this File is not
+	// open.
```
Comment (schomatis) [inline file.go:171]: I agree with the logic explained here but I think we should better document `desclock` instead. The vague 'coordination' should explicitly say: this lock makes sure there is either only one opened write-mode file descriptor or many read-mode ones.

File: mfs_test.go [hunk 6]
```diff
@@ -246,6 +246,25 @@ func TestBasic(t *testing.T) {
...
+		t.Fatal("timed out waiting for Close")
+	}
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	err = rt.Close()
+	if err != nil {
+		t.Fatal(err)
+	}
```
Comment (schomatis) [inline mfs_test.go:267]: I'm assuming this is testing that we can close an already closed `Root` with no error and no side-effect, if so please document (as `(*Root).Close()` has no doc).

**Review by schomatis — COMMENTED** [review]
Review in progress, publishing some notes in the meanwhile. In the future please try to decouple logically independent patches (edit: particularly any change in the `Republisher` should have its own PR, or commit at least).

File: dir.go (hunk 1 above)
Comment (gammazero) [reply dir.go:125]: The `dagService`. [adds the block])https://github.com/ipfs/go-merkledag/blob/master/merkledag.go#L51) to the blockstore,  The blockservice [puts](https://github.com/ipfs/go-blockservice/blob/master/blockservice.go#L146) the block into the blockstore.  The blocksstore [puts](https://github.com/ipfs/go-ipfs-blockstore/blob/master/blockstore.go#L169) the raw block data into the datastore.  This [raw data](https://github.com/ipfs/go-block-format/blob/master/blocks.go#L64) is a byte slice from the block, not a copy, and the memory it references leads to a race condition.  I am not sure where that memory gets overwritten, but running tests with the race detector will flag the race condition.

File: dir.go (hunk 2 above)
Comment (gammazero) [reply dir.go:409]: If there is an error in the loop, then the function will return and not clear the cache entries.  If possible these can be retried later.  Only if the loop completes successfully are the cache entries cleared, which is safe because all the entries have been updated.  This avoids unnecessary updates in the future due to nodes that are still in the cache but are already up-to-date.

File: file.go (hunk 4 above)
Comment (gammazero) [reply file.go:168]: Maybe,  but since this function is exported it could be called externally or used to satisfy some interface somewhere.  I did not want to introduce any potential breakage in this PR.

File: mfs_test.go (hunk 6 above)
Comment (gammazero) [reply mfs_test.go:269]: done

File: file.go (hunk 5 above)
Comment (gammazero) [reply file.go]: Updated comment to explain in more detail.

File: repub.go [hunk 7]
```diff
@@ -167,23 +176,14 @@ func (rp *Republisher) Run(lastPublished cid.Cid) {
...
-				case <-time.After(rp.RetryTimeout):
-				case <-rp.ctx.Done():
-					return
-				}
+			err := rp.pubfunc(ctx, toPublish)
+			if err != nil {
+				// Keep retrying until publish succeeds, or run is stopped.
+				// Publish newer values if available.
+				longer.Reset(timeoutLong)
+				continue
```
Comment (schomatis) [inline repub.go:186]: It seems we now don't "keep retrying" (we should reword that) and instead move on to the next loop (whether a new value to publish comes along superseding the failed one is not relevant). I would then suggest we refactor the next `WaitPub` (third) step into this one.

Before, the 1-2-3 succession implied all of them were always going to happen (at most step 3 would be delayed until the publish succeeded). Steps 2-3 should just be a single block now handling the fallout from the publish call: if successful, update `lastPublished` and trigger `waiter`, if not, do nothing and move on to the next loop (fine if that is not an actual big `if` and we keep the `continue` but it should more clearly reflect this).

File: repub_test.go [hunk 8]
```diff
@@ -14,12 +14,14 @@ func TestRepublisher(t *testing.T) {
...
-
 	pub := make(chan struct{})
 
 	pf := func(ctx context.Context, c cid.Cid) error {
-		pub <- struct{}{}
+		select {
+		case pub <- struct{}{}:
+		case <-ctx.Done():
+			return ctx.Err()
+		}
```
Comment (schomatis) [inline repub_test.go:24]: This, along with the way `ctx` is passed as an argument now instead of storing it in the `Republisher`, makes much more sense.

File: repub_test.go [hunk 9]
```diff
@@ -29,8 +31,10 @@ func TestRepublisher(t *testing.T) {
 	tshort := time.Millisecond * 50
 	tlong := time.Second / 2
 
-	rp := NewRepublisher(ctx, pf, tshort, tlong)
-	go rp.Run(cid.Undef)
+	ctx, cancel := context.WithTimeout(context.Background(), 30*time.Second)
+	defer cancel()
```
Comment (schomatis) [inline repub_test.go:35]: [Question for my own edification] Is this `defer cancel()` call just good old clean Go code (which I'm in favor of) or is it actually preventing some goroutine from continuing to run after the tests ends? (I always assume the test will kill everything when it ends but that may not be the case.)

File: repub.go [hunk 10]
```diff
@@ -13,36 +14,35 @@ type PubFunc func(context.Context, cid.Cid) error
...
-		TimeoutShort:     tshort,
-		TimeoutLong:      tlong,
-		RetryTimeout:     tlong,
+	rp := &Republisher{
 		update:           make(chan cid.Cid, 1),
 		pubfunc:          pf,
 		immediatePublish: make(chan chan struct{}),
-		ctx:              ctx,
 		cancel:           cancel,
+		stopped:          make(chan struct{}),
```
[the remark this answers is not in the document]
Comment (schomatis) [reply repub.go:36]: This was actually more straightforward than I originally thought, but still it would be useful to add a concise comment explaining this channel will be closed after the main `run` goroutine ends (considering now that `run` is triggered internally only once and not by the client).

File: repub.go [hunk 11]
```diff
@@ -58,46 +58,57 @@ func (rp *Republisher) WaitPub(ctx context.Context) error {
...
 }
 
-func (rp *Republisher) Close() error {
-	// TODO(steb): Wait for `Run` to stop
-	err := rp.WaitPub(rp.ctx)
-	rp.cancel()
-	return err
+// Close tells the republisher to stop and waits for it to stop.  If it is
+// necessary to wait for any current values to be published, then call WaitPub
+// prior to calling Close.
```
Comment (schomatis) [inline repub.go:63]: I get the logic but I'm not sure I understand this sentence. Are we suggesting the user to call `WaitPub` or is this a description of what we are doing internally?

File: repub_test.go [hunk 12]
```diff
@@ -66,10 +69,78 @@ func TestRepublisher(t *testing.T) {
...
+	go func() {
+		defer close(stopped)
+		for {
+			select {
+			case <-pub:
+			case <-stopPub:
+				return
+			}
+		}
+	}()
```
Comment (schomatis) [inline repub_test.go:107]: I'm not following the objective here, why do we need this? At this point it seems we don't interact with `pub` anymore (and we stopped the continuous updates with `stopUpdates`), if we need a publish function that doesn't block we should redefine it here to make this more explicit and reduce complexity (but maybe I'm missing the point and that was not the use of this goroutine).

**Review by schomatis — COMMENTED** [review]
Notes on the `Republisher`. I'm still missing the `Root` fixes.

I like the documentation improvements. They are also a good candidate to encapsulate in a separate commit to make it easier to spot where is the actual logic being changed (to which I need to pay more attention) and where it's not.

File: dir.go (hunk 2 above)
Comment (schomatis) [reply dir.go:409]: I get the rationale of why do we clear the cache in general, my question was more specific to the case where, say, I have 2 entries in the cache, `A` and `B`, then the loop processes correctly `A` and fails at `B`, but the cache still has both `A` and `B` in it because we exited early inside the loop (when `updateChild` returned an error) and we didn't reach this line. The crux of my question then is: a future call to `sync` will update again `A` which we didn't want, potentially overwriting a newer entry with same name, is this an issue? If so, do we want to fix it, add a `FIXME`, comment it as part of the API?

File: file.go (hunk 4 above)
Comment (schomatis) [reply file.go:168]: Fair enough. Usually `go-ipfs` is the main consumer here but we should check more broadly (@Stebalien ?) and clean in a follow-up PR.

File: file.go (hunk 5 above)
Comment (schomatis) [reply file.go]: I'd rather have this kind of explanation in the more central `desclock` definition than inside a function manipulation but it's good as it is so I consider this fixed.

File: dir.go (hunk 1 above)
Comment (schomatis) [reply dir.go:125]: I agree with your assessment of the block data model but I still fail to see where is the potential race.

By this point in the code `dagService.Add()` has finished and the component that has access to this byte slice (`BlockService`) has already stored it in the `Datastore`. These components AFAIK do not keep a reference to the slice; in a way _they_ are the ones doing the copying (actually dumping the bytes in the store medium) of the slice, so why do we need to copy as well? Could you point me to the source line where you think the access might cause the race please?

File: repub_test.go (hunk 9 above)
Comment (gammazero) [reply repub_test.go:35]: It is boilerplate good practice so that if the test exited unexpectedly it would allow any goroutines waiting on the context to exit.  If we add add goroutine leak detection at some point, this kind of thing also helps prevent false positives.

File: file.go (hunk 4 above)
Comment (Stebalien) [reply file.go:168]: Go-ipfs is pretty much the only consumer.

File: root.go [hunk 13]
```diff
@@ -204,14 +193,22 @@ func (kr *Root) updateChildEntry(c child) error {
...
+	defer dir.lock.Unlock()
+
 	if kr.repub != nil {
-		kr.repub.Update(nd.Cid())
-		return kr.repub.Close()
+		repub := kr.repub
+		kr.repub = nil
+		repub.Update(nd.Cid())
+		return repub.Close(kr.ctx)
 	}
```
Comment (schomatis) [inline root.go:212]: [I'm missing some context here and the change is definitely an improvement so feel free to ignore.] Why do we update in this case? I understand flushing in the `Close` of the file descriptor but I'm not sure I want to publish changes made directly to this structure (I'm not even sure who would manipulate the `Root` directly and even need a `Close` API except to stop the republisher).

**Review by schomatis — APPROVED** [review]
`Root` looks good. This completes my review. Everything looks good, just some minor comments that I'd like addressed, but nothing major so feel free to merge if this is time sensitive.

File: dir.go (hunk 2 above)
Comment (gammazero) [reply dir.go:414]: As far as I can tell, a future call to sync in the case described above is harmless, and will only cause an unnecessary update of the same data that is already present.  It will not overwrite a newer entry with the same name, as the creation/update of the newer entry will also update the cache.

However, in the case of an unnecessary update following a sync failure, the timestamp of the entry may get updated.  So, I now delete each entry from the cache as soon as it is synced.

File: file.go (hunk 4 above)
Comment (gammazero) [reply file.go]: Removed.  Cannot find any use of it.

File: repub.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (gammazero) [reply repub.go:36]: Added comment in type definition.

File: repub.go (hunk 7 above)
Comment (gammazero) [reply repub.go:192]: The logic does retry in that it resets the timer and continues the loop, retrying the pubfunc on the next timer expiration.  The previously failed value is retried unless a new value is provided.  Updated the comments to clarify.

Fixed the problem where this could lose a waiter if there were multiple calls to `WaitPublish()` during retry.  Added unit test for this.

File: repub.go (hunk 11 above)
Comment (gammazero) [reply repub.go]: It is saying that the user should not expect `WaitPub()` to work after calling `Close()`, so if they need to use `WaitPub()` they need to finish doing that before calling `Close()`. I think that goes without saying, so I removed the comment.

File: repub_test.go (hunk 12 above)
Comment (gammazero) [reply repub_test.go]: This is necessary to continue clearing the `pub` channel.  Otherwise, the republisher's call to `pubfunc` will never return, and calling `Close()` will timeout.  The `stopped` channel is however not needed.

File: root.go (hunk 13 above)
Comment (gammazero) [reply root.go]: I am guessing is needed to ensure the root CID is published in the case where there were no updates to MFS.

As far as needing the Close() API... Part of graceful shutdown perhaps.  Did not intend to analyze that potential call path.  Just wanted to maintain functional equivalence here with safer code.

File: dir.go (hunk 1 above)
Comment (gammazero) [reply dir.go:125]: It is possible the race condition is only induced by the test code.  I will take more time to see if this is possible outside of the test.

**Comment by gammazero** [comment]
@schomatis I think I have addressed all review comments if you would like to have a final look.

**Review by schomatis — APPROVED** [review]
Thanks for addressing all my nitpicks. I picked up a few nice concurrency patterns along this review.

File: dir.go [hunk 14]
```diff
@@ -403,9 +403,15 @@ func (d *Directory) sync() error {
 		if err != nil {
 			return err
 		}
+
+		// Although cache is cleared at end of function, delete each entry
+		// after update to prevent unnecessary future update in case of failure
+		// before end of loop
+		delete(d.entriesCache, name)
```
Comment (Stebalien) [inline dir.go:410]: This is unsafe, IIRC, if we have open files. That's why we have a separate "this is totally unsafe only call this if you know what you're doing" `Uncache` function.

TL;DR: the cache isn't a cache, it's a record of open files.

File: file.go [hunk 15]
```diff
@@ -71,17 +73,16 @@ func (fi *File) Open(flags Flags) (_ FileDescriptor, _retErr error) {
...
+	node, err := fi.GetNode()
+	if err != nil {
+		return nil, err
+	}
 
-	// TODO: Move this `switch` logic outside (maybe even
-	// to another package, this seems like a job of UnixFS),
-	// `NewDagModifier` uses the IPLD node, we're not
+	// TODO: Move this logic outside (maybe even to another package, this seems
+	// like a job of UnixFS), `NewDagModifier` uses the IPLD node, we're not
```
Comment (Stebalien) [inline file.go:82]: I think the only reason we're doing this is because we don't support symlinks... We should just open the file then check if it's a symlink.

**Review by Stebalien — CHANGES_REQUESTED** [review]
Are we fixing any known bugs here? This code is prone to explode at the slightest touch so I'd rather not refactor little parts of it just to address some TODO items.

File: dir.go (hunk 14 above)
Comment (schomatis) [reply dir.go:410]: _Please provide more context and expand on your argument._

We are already clearing the cache after the loop. Is _this_ particularity line 'unsafe', or any cache modification inside this function?

> the cache isn't a cache, it's a record of open files.

What does this mean? This structure functions like a cache. The `Uncache` function is just an ugly optimization to prevent cache bloating: https://github.com/ipfs/go-ipfs/pull/3027#issuecomment-448678709. `Uncache` deletes a cache entry, same as `Unlink` (we won't flush an entry we are removing) and `sync` here.

Comment (Stebalien) [reply dir.go:410]: > We are already clearing the cache after the loop. Is this particularity line 'unsafe', or any cache modification inside this function?

Where? As far as I can tell, this PR _adds_ the cache clearing after the loop (which has the same problem).

> What does this mean? This structure functions like a cache.

https://github.com/ipfs/go-mfs/issues/54?

File: dir.go (hunk 1 above)
Comment (schomatis) [reply dir.go:125]: `localUpdate` here is an internal function. We are not exposing the returned internal structure in the API. As its documentation describes, `localUpdate` is just an extraction from the bigger (also internal) `updateChildEntry`.

File: file.go (hunk 15 above)
Comment (schomatis) [reply file.go:82]: There is no logic change here. Just a reformat of an old comment.

File: dir.go (hunk 14 above)
Comment (schomatis) [reply dir.go:410]: Yes, we've been abusing this. I still think removing the entry here is no more evil than what is happening elsewhere, but if this is just a performance optimization we can skip it.

Comment (gammazero) [reply dir.go]: It is an attempt to reduce memory use, not so much for performance.  It may actually decrease performance by having to retrieve a node from the DAG, instead of getting it from the cache.

After the above discussion, I have stopped clearing `entriesCache` in `sync()`.

File: file.go (hunk 15 above)
Comment (gammazero) [reply file.go:82]: @Stebalien This TODO, and the TODO below it, will probably disappear when we move to use an IPLD-prime node.  So, will leave this as it for now.

File: dir.go (hunk 14 above)
Comment (Stebalien) [reply dir.go]: > Yes, we've been abusing this. I still think removing the entry here is no more evil than what is happening elsewhere, but if this is just a performance optimization we can skip it.

It's bad, but there we _know_ it's safe because we own the MFS object. Here, it's definitely unsafe (concurrent `ipfs files *` operations will race and conflict).

**Comment by gammazero** [comment]
> Are we fixing any known bugs here? This code is prone to explode at the slightest touch so I'd rather not refactor little parts of it just to address some TODO items.

The short answer is no, this PR takes #58 a little farther by fixing up some of the TODO items and fixing a potential race condition noted [here](https://github.com/ipfs/go-mfs/pull/85/files#diff-6e3e18cb7605f7bcf8be86cf21f3698516ef4bcd257fa21f31a646cf7cc068f3L197-L198).  The other reason for these changes is in preparation for modifications to use ipld-prime -- to clean up anything that can be cleaned up first.

If these changes are not worth it in your opinion, then I will close PR but keep this branch (for now) to use as a reference for later.

**Comment by Stebalien** [comment]
Fixing race conditions is good. I'll take another look at the PR.

> The other reason for these changes is in preparation for modifications to use ipld-prime -- to clean up anything that can be cleaned up first.

The current library is fundamentally broken at the API level due to issues like #54. I'm not convinced fixing this library is worth it, rather than rewriting.

**Comment by aschmahmann** [comment]
@Stebalien will take a look at this this week and if we're able to get a few clean bugfixes we'll merge otherwise we'll close awaiting a deeper refactor/rewrite.

**Comment by aschmahmann** [comment]
closing awaiting a deeper refactor
